common/streams/src/sources/source_parquet.rs

```diff
@@ -31,141 +30,94 @@ use common_exception::Result;
+                match self.metadata.as_ref() {
+                    Some(m) => m,
+                    _ => unreachable!(),
+                }
             }
         };
 
-        self.row_groups = metadata.row_groups.len();
-        self.row_group = 0;
-
-        if self.row_group >= self.row_groups {
+        if self.current_row_group >= metadata.row_groups.len() {
             return Ok(None);
         }
-        let col_num = self.projection.len();
-        let row_group = self.row_group;
+
+        let fields = self.arrow_table_schema.fields();
+        let row_grp = &metadata.row_groups[self.current_row_group];
         let cols = self
             .projection
             .clone()
             .into_iter()
-            .map(|idx| (metadata.row_groups[row_group].column(idx).clone(), idx));
-
-        let fields = self.arrow_table_schema.fields();
-        let stream_len = self.file_len;
-        let read_buffer_size = self.read_buffer_size.unwrap_or(DEFAULT_READ_BUFFER_SIZE);
-
-        let stream = futures::stream::iter(cols).map(|(col_meta, idx)| {
-            let data_accessor = self.data_accessor.clone();
-            let path = self.path.clone();
-
-            async move {
-                let reader = data_accessor.get_input_stream(path.as_str(), stream_len)?;
-                let mut reader = BufReader::with_capacity(read_buffer_size as usize, reader);
-                // TODO cache block column
-                let col_pages =
-                    get_page_stream(&col_meta, &mut reader, vec![], Arc::new(|_, _| true))
-                        .instrument(debug_span!("parquet_source_get_column_page"))
-                        .await
-                        .map_err(|e| ErrorCode::ParquetError(e.to_string()))?;
-                let pages = col_pages.map(|compressed_page| {
-                    debug_span!("parquet_source_decompress_page")
-                        .in_scope(|| decompress(compressed_page?, &mut vec![]))
-                });
-                let array = page_stream_to_array(pages, &col_meta, fields[idx].data_type.clone())
-                    .instrument(debug_span!("parquet_source_page_stream_to_array"))
-                    .await?;
-                let array: Arc<dyn common_arrow::arrow::array::Array> = array.into();
-                Ok::<_, ErrorCode>(DataColumn::Array(array.into_series()))
-            }
-            .instrument(debug_span!("parquet_source_read_column").or_current())
-        });
-
-        // TODO configuration of the buffer size
-        let buffer_size = 10;
-        let n = std::cmp::min(buffer_size, col_num);
-        let data_cols = stream.buffered(n).try_collect().await?;
-
-        self.row_group += 1;
+            .map(|idx| (row_grp.column(idx).clone(), idx));
+        let mut data_cols = Vec::with_capacity(cols.len());
+        for (col_meta, idx) in cols {
+            let col_pages =
+                get_page_stream(&col_meta, &mut self.reader, vec![], Arc::new(|_, _| true))
```

Comment:
So we do not read columns in parallel, will it slow down the performance?

Reply:
for fuse engine, the newly added `BlockReader` will be used, which keeps reading the columns parallelly.